nixos/redshift: rework to use config file

hyperfekt opened the pull request:
###### Motivation for this change
Redshift has some more options that are not reachable via the CLI (for example dawn-time and dusk-time) - this inspired me to rework the module to allow configuring it via the config file directly, automatically allowing users to take advantage of any options redshift might offer with a convenient (although less directly documented) interface while decreasing the burden of maintaining the module.

This is more of a proposal and not yet tested at all, I'd like to gather some input if this model is something that has a chance of getting merged at all, if there are any big improvements that could be made to it, any big problems I've overlooked etc.
Specifically I'd be interested in adding back some level of in-line documentation, implementing backwards-compatibility if wished and the most elegant way to check if an attribute is set (for the asserts) - that last one might interact with the first one; I've just put a placeholder for now, checking for null would obviously throw errors.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -93,52 +34,59 @@ in {
+          }
+          {
+            names = "Time of dawn and time of dusk";
+            items = times;
```
Do these really need to be provided together?

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -93,52 +34,59 @@ in {
+            message = "The transitions must be set via exactly one mechanism.";
+          }
+          {
+            assertion = if manualLocationProvider then (allThe coordinates) else (no coordinates);
```
Just do this

```nix
{
  assertion = cfg.config.redshift.location-provider == "manual" ->
    cfg.config ? manual.lon && cfg.config ? manual.lat;
  message = "services.redshift.config.manual.{lon,lat} need to be provided if the location provider is \"manual\";
}
```

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -93,52 +34,59 @@ in {
+        [
+          {
+            assertion = oneOf mechanisms;
+            message = "The transitions must be set via exactly one mechanism.";
```
What happens if both are set? This assertion seems a bit unnecessary

hyperfekt replied:
Nothing bad, the manually set times will take precedence; but I thought if we had a chance to only allow configurations that follow the user's expectations and disallow ambiguous ones where we could that'd be a win.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -93,52 +34,59 @@ in {
-            -t ${toString cfg.temperature.day}:${toString cfg.temperature.night} \
-            -b ${toString cfg.brightness.day}:${toString cfg.brightness.night} \
-            ${lib.strings.concatStringsSep " " cfg.extraOptions}
+            -c ${lib.toINI cfg.config}
```
`lib` doesn't have a `toINI`, did you even evaluate this? It's in `lib.generators.toINI`

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -18,73 +18,14 @@ in {
-          between <literal>0.1</literal> and <literal>1.0</literal>.
-        '';
-      };
-      night = mkOption {
```
You would have to introduce backwards compatible aliases for these removed options with `mkAliasOptionModule`

hyperfekt replied:
I've chosen mkRenamedOptionModule in order to get people to migrate to the new configuration method in case the config file options change in a way that doesn't allow an alias in the future.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -18,73 +18,14 @@ in {
-        Your current latitude, between
-        <literal>-90.0</literal> and <literal>90.0</literal>. Must be provided
-        along with longitude.
+        Configuration to give to redshift.
```
The description should be a lot longer, explain what this option does and how it can be used.

hyperfekt replied:
Referring to man page now, I hope that's sufficient.

infinisil reviewed (changes requested):
While I like the `config` option in general, I'm not sure if it's fit for redshift, because there's not too many things you can configure anyways, and I can't find any docs on what options are supported in general (link me to it if there is). This combination makes it not very attractive to use such an unchecked, undocumented `config` option. Let me know if you have a good reason for it.

I introduced the same thing for https://github.com/NixOS/nixpkgs/pull/45470, but there ZNC has a lot of options, which are all documented well, and there are too many for NixOS, and they change a lot in addition, which wouldn't be very maintainable.

hyperfekt commented:
> I can't find any docs on what options are supported in general
You're correct, unfortunately redshift does not provide any documentation for its config file. What do you think about leaving the existing options in place, but exposing the config file as an option (and realizing the existing ones through that) in order to allow composition for any settings defined therein?

hyperfekt commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -18,73 +18,14 @@ in {
-      type = types.nullOr types.str;
-      default = null;
+    config = mkOption {
+      type = types.attrs;
```
If I'm not gravely mistaken merging works fine for attrs, but the INI format used by redshift does has a specific structure that should be reflected in the type, so I'll adopt this.

Infinisil commented:
One of the main motivations for the `config` option for ZNC was that there isn't the need anymore to maintain docs in NixOS. Docs should be part of the package and if redshift doesn't have that then it should be fixed on the redshift side imo.

hyperfekt commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -93,52 +34,59 @@ in {
+      with lib; let
+        allThe = x: !(any (map isNull x));
+        no = x: all (map isNull x);
+        oneOf = xs: (count id xs) == 1;
```
Now that the method of inspection is determined, I've the first two functions; and I've inlined the last one.

hyperfekt commented:
It turns out that redshift, contrary to several statements on the web, actually has a complete documentation for the config file: It can be found on the man page.
As a result, I've referred to that as documentation for the .config option and gone ahead with the change without providing any further documentation for individual options.
If you'd still like me to reduce the abstractions in the assertions, I can do that.

hyperfekt commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -93,52 +34,59 @@ in {
+        (map together [ 
+          {
+            names = "Latitude and longitude";
+            items = coordinates;
```
The tradeoff in integrating this is that the error message becomes both longer and less accurate.
Given that we need something like this anyway for `dusk-time` and `dawn-time`, would you still like it removed?

xaverdh commented:
Just noticed this pull request.
You may be interested in #57554.

Infinisil commented:
Btw I've since opened an RFC for such `config` options: https://github.com/NixOS/rfcs/pull/42

hyperfekt commented:
Wow, I only just noticed I never removed the [WIP] tag. -_-

hyperfekt commented:
This is ultimately up to @Infinisil (and the RFC, I guess); but the way @xaverdh did it in #57975 a collision between a module-provided option and a config attribute set option will have the latter silently overwrite the former instead of failing like I think it should.
I've integrated the earlier target for the drm method here, I wasn't sure how to test that but I assume @xaverdh did that.

Infinisil commented:
https://github.com/NixOS/nixpkgs/pull/64309 is very relevant for this and almost in shape to be merged, this PR will need some adjustments after that. Pinging @eadwu as well

Commit:
nixos/redshift: rework to use config files

also uses an earlier target for the drm adjustment method,
and allows using dawn / dusk time even if a location provider is set

hyperfekt commented:
Adjusted accordingly.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -35,42 +46,41 @@ in {
+      apply = c:
+        if !(c ? redshift.dawn-time || c ? redshift.dusk-time) && !(c ? redshift.location-provider) && options.locations.provider.isDefined then
+          c // {
+            redshift.location-provider = lcfg.provider;
```
This clears all other `redshift.*` keys, you need to use `recursiveUpdate` instead

Infinisil replied:
And `location.provider` is always defined because it has a default, also `c ? dawn-time == c ? dusk-time` as per assertion, so this can be
```nix
c: if ! c ? redshift.dawn-time && ! c ? redshift.location-provider
  then recursiveUpdate c {
    redshift.location-provider = lcfg.provider;
  } else c
```

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -82,18 +92,21 @@ in {
+    } ];
+
+    services.redshift.settings.manual = {
+      lat = mkIf options.location.latitude.isDefined lcfg.latitude;
```
I think this should be something like
```nix
let
  needsLocationProvider = ! cfg.settings ? redshift.dawn-time;
in mkIf (needsLocationProvider && lcfg.provider == "manual") {
  lat = lcfg.latitude;
  lon = lcfg.longitude;
}
```

Setting `lcfg == "manual"` should force the user to set `longitude` and `latitude` if it's needed. `isDefined` doesn't do that.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -82,18 +92,21 @@ in {
+
+    assertions = mkIf (cfg.configFile == generatedConfig) [ {
+        assertion = (cfg.settings ? redshift.dawn-time) == (cfg.settings ? redshift.dusk-time);
+        message = "Time of dawn and time of dusk must be provided together.";
```
The assertion message should also mention that this is about redshift

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -82,18 +92,21 @@ in {
+    services.redshift.configFile = mkDefault generatedConfig;
+
+    assertions = mkIf (cfg.configFile == generatedConfig) [ {
+        assertion = (cfg.settings ? redshift.dawn-time) == (cfg.settings ? redshift.dusk-time);
```
No parens needed

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -35,42 +46,41 @@ in {
-        '';
-      };
+    settings = mkOption {
+      type = with types; attrsOf (attrsOf (nullOr (oneOf [ str float bool int ])));
```
`toINI` doesn't seem to support floats yet. And you should either remove `nullOr` or make sure `null` results in nothing being generated.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -35,42 +46,41 @@ in {
+      default = {};
+      description = ''
+        The configuration to pass to redshift.
+        See <command>man redshift</command> under the section
```
This is the nasty syntax for man pages: `<citerefentry><refentrytitle>redshift</refentrytitle><manvolnum>1</manvolnum></citerefentry>`

Valodim commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -35,42 +46,41 @@ in {
+        <option>settings</option> option instead.
+        </para>
+        <para>
+        Setting this option will override any configuration file auto-generated
```
Shouldn't this be disallowed by assertion? Allowing a `settings` object that's going to have no effect since it's overridden adds potential for confusion but little value, doesn't it?

xaverdh commented:
@hyperfekt are you still working on this?
(people started implementing this again in #108625)

Lassulus commented:
seems like the other PRs were abandoned. Do you want to continue this?
